File: session-2/402.java
/*
 * 402. Remove K Digits
 */
public class Solution {
    private String removeOneDigit(String num) {
        if (num.length() == 1) return "0";
        if (num.length() >= 2 && num.charAt(1) == '0') {
            int idx = 1;
            while (idx < num.length() && num.charAt(idx) == '0')
                idx++;
            String res = num.substring(idx);
            if (res.equals("")) return "0";
            return res;
        }

        int i = 0;
        while (i < num.length() - 1 && num.charAt(i) <= num.charAt(i+1))
            i++;
        return num.substring(0, i) + num.substring(i + 1);
    }

    public String removeKdigits(String num, int k) {
        if (num.length() == k) return "0";
        while (k > 0) {
            num = removeOneDigit(num);
            k--;
        }
        return num;
    }
}

File: session-2/402.cpp
// 402. Remove K Digits
#include "leetcode.hpp"
using namespace std;

class Solution {
public:
    string removeKdigits(string num, int k) {        
        stack<char> st;
        for (int i = 0; i < num.size(); ++i) {
            char c = num[i];
            while (st.size() && st.top() > c && k > 0) {
                st.pop();
                k -= 1;
            }
            st.push(c);
        }

        while (k > 0 && st.size()) {
            st.pop();
            k--;
        }

        string result;
        while (st.size()) {
            result.push_back(st.top());
            st.pop();
        }
        reverse(result.begin(), result.end());

        int begin = 0;
        while (begin < result.size() && result[begin] == '0') begin++;
        if (begin >= result.size()) return "0";

        return result.substr(begin);
    }
};

int main() {
    Solution s;
    cout << s.removeKdigits("10", 3) << endl;
}